test/common/network/listen_socket_impl_test.cc

```diff
@@ -2,43 +2,52 @@
 
 #include "common/network/utility.h"
 #include "common/network/listen_socket_impl.h"
+#include "test/test_common/network_utility.h"
 
 namespace Network {
 
-TEST(ListenSocket, All) {
-  // Test the case of a socket with given port and bind_to_port set to true.
-  TcpListenSocket socket1(uint32_t(15000), true);
+class ListenSocketImplTest : public testing::TestWithParam<Address::IpVersion> {
+protected:
+  ListenSocketImplTest() : version_(GetParam()) {}
+  const Address::IpVersion version_;
+};
+INSTANTIATE_TEST_CASE_P(IpVersions, ListenSocketImplTest,
+                        testing::Values(Address::IpVersion::v4, Address::IpVersion::v6));
+
+TEST_P(ListenSocketImplTest, BindSpecificPort) {
+  // Pick a free port.
+  auto addr_fd = Network::Test::bindFreeLoopbackPort(version_, Address::SocketType::Stream);
+  auto addr = addr_fd.first;
+  EXPECT_LE(0, addr_fd.second);
+
+  // Confirm that we got a reasonable address and port.
+  ASSERT_EQ(Address::Type::Ip, addr->type());
+  ASSERT_EQ(version_, addr->ip()->version());
+  ASSERT_LT(0U, addr->ip()->port());
+
+  // Release the socket and re-bind it.
```

Comment:
I'm still a little unhappy with this test for IPv6 and the flake potential due to a single IPv6 loopback address being available. Can we make this IPv4-only or have it do something like retry a bounded number of times (100?) on failure until it works?